app/featuresmodel.cpp

```diff
@@ -25,31 +35,77 @@ void FeaturesModel::populate()
 {
   if ( mLayer )
   {
+    mFetchingResults = true;
+    emit fetchingResultsChanged();
     beginResetModel();
     mFeatures.clear();
+    endResetModel();
 
     QgsFeatureRequest req;
     setupFeatureRequest( req );
 
-    QgsFeatureIterator it = mLayer->getFeatures( req );
-    QgsFeature f;
+    int searchId = mNextSearchId.fetchAndAddOrdered( 1 );
+    QgsVectorLayer *layer = mLayer->clone();
+    layer->moveToThread( nullptr );
+    QFuture<QgsFeatureList> future = QtConcurrent::run( &FeaturesModel::fetchFeatures, this, layer, req, searchId );
+    mSearchResultWatcher.setFuture( future );
+  }
+}
 
-    while ( it.nextFeature( f ) )
+QgsFeatureList FeaturesModel::fetchFeatures( QgsVectorLayer *layer, QgsFeatureRequest req, int searchId )
```

Comment:
It would be better to have fetchFeatures() a simple static function outside of `FeaturesModel`, we would only pass it reference to `mNextSearchId` - better to have least amount of shared data in a function that runs in a worker thread.

Reply:
This gave me a hard time and I gave up. As soon as I add `QAtomicInt &` to the signature I can't get it to compile...
<details>
  <summary>sad compiler</summary>
```
In file included from /usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtaskbuilder.h:9,
                 from /usr/include/x86_64-linux-gnu/qt6/QtConcurrent/QtConcurrent:4,
                 from /home/uclaros/dev/mobile/input/app/featuresmodel.cpp:19:
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h: In instantiation of ‘struct QtConcurrent::PromiseTaskResolver<QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>’:
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:211:8:   required from ‘struct QtConcurrent::TaskResolverHelper<std::integral_constant<bool, false>, QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:217:8:   required from ‘struct QtConcurrent::TaskResolver<QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentrun.h:40:76:   required from ‘auto QtConcurrent::run(QThreadPool*, Function&&, Args&& ...) [with Function = QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&); Args = {FeaturesModel*, QgsVectorLayerFeatureSource*&, QgsFeatureRequest&, int&, QAtomicInt&}]’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentrun.h:57:15:   required from ‘auto QtConcurrent::run(Function&&, Args&& ...) [with Function = QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&); Args = {FeaturesModel*, QgsVectorLayerFeatureSource*&, QgsFeatureRequest&, int&, QAtomicInt&}]’
/home/uclaros/dev/mobile/input/app/featuresmodel.cpp:46:54:   required from here
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:189:64: error: static assertion failed: The first argument of passed callable object isn't a QPromise<T> & type. Did you intend to pass a callable which takes a QPromise<T> & type as a first argument? Otherwise it's not possible to invoke the function with passed arguments.
  189 |     static_assert(QtPrivate::ArgResolver<Function>::IsPromise::value,
      |                                                                ^~~~~
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:189:64: note: ‘std::integral_constant<bool, false>::value’ evaluates to false
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h: In instantiation of ‘struct QtConcurrent::MemberFunctionResolver<QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), void, FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>’:
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:77:8:   required from ‘struct QtConcurrent::FunctionResolverHelper<std::integral_constant<bool, true>, QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), void, FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:83:8:   required from ‘struct QtConcurrent::FunctionResolver<QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), void, FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:140:11:   required from ‘struct QtConcurrent::StoredFunctionCallWithPromise<QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), void, FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:196:17:   required from ‘static auto QtConcurrent::PromiseTaskResolver<Function, Args ...>::run(TaskWithArgs&&, const QtConcurrent::TaskStartParameters&) [with Function = QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&); Args = {FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt}; TaskWithArgs = std::tuple<QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>]’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentrun.h:40:76:   required from ‘auto QtConcurrent::run(QThreadPool*, Function&&, Args&& ...) [with Function = QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&); Args = {FeaturesModel*, QgsVectorLayerFeatureSource*&, QgsFeatureRequest&, int&, QAtomicInt&}]’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentrun.h:57:15:   required from ‘auto QtConcurrent::run(Function&&, Args&& ...) [with Function = QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&); Args = {FeaturesModel*, QgsVectorLayerFeatureSource*&, QgsFeatureRequest&, int&, QAtomicInt&}]’
/home/uclaros/dev/mobile/input/app/featuresmodel.cpp:46:54:   required from here
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:51:24: error: static assertion failed: It's not possible to invoke the function with passed arguments.
   51 |     static_assert(std::is_invocable_v<std::decay_t<Function>, std::decay_t<Arg>, QPromise<PromiseType> &, std::decay_t<Args>...>,
      |                   ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:51:24: note: ‘std::is_invocable_v<QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), FeaturesModel*, QPromise<void>&, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>’ evaluates to false
In file included from /usr/include/x86_64-linux-gnu/qt6/QtCore/qglobal.h:9,
                 from /usr/include/x86_64-linux-gnu/qt6/QtCore/qcontainertools_impl.h:14,
                 from /usr/include/x86_64-linux-gnu/qt6/QtCore/qhash.h:8,
                 from /usr/include/x86_64-linux-gnu/qt6/QtCore/qabstractitemmodel.h:8,
                 from /usr/include/x86_64-linux-gnu/qt6/QtCore/QAbstractListModel:1,
                 from /home/uclaros/dev/mobile/input/app/featuresmodel.h:13,
                 from /home/uclaros/dev/mobile/input/app/featuresmodel.cpp:10:
/usr/include/c++/12/type_traits: In substitution of ‘template<class _Fn, class ... _Args> using invoke_result_t = typename std::invoke_result::type [with _Fn = QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&); _Args = {FeaturesModel*, QPromise<void>&, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt}]’:
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:53:24:   required from ‘struct QtConcurrent::MemberFunctionResolver<QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), void, FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:77:8:   required from ‘struct QtConcurrent::FunctionResolverHelper<std::integral_constant<bool, true>, QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), void, FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:83:8:   required from ‘struct QtConcurrent::FunctionResolver<QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), void, FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:140:11:   required from ‘struct QtConcurrent::StoredFunctionCallWithPromise<QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), void, FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentstoredfunctioncall.h:196:17:   required from ‘static auto QtConcurrent::PromiseTaskResolver<Function, Args ...>::run(TaskWithArgs&&, const QtConcurrent::TaskStartParameters&) [with Function = QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&); Args = {FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt}; TaskWithArgs = std::tuple<QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), FeaturesModel*, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>]’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentrun.h:40:76:   required from ‘auto QtConcurrent::run(QThreadPool*, Function&&, Args&& ...) [with Function = QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&); Args = {FeaturesModel*, QgsVectorLayerFeatureSource*&, QgsFeatureRequest&, int&, QAtomicInt&}]’
/usr/include/x86_64-linux-gnu/qt6/QtConcurrent/qtconcurrentrun.h:57:15:   required from ‘auto QtConcurrent::run(Function&&, Args&& ...) [with Function = QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&); Args = {FeaturesModel*, QgsVectorLayerFeatureSource*&, QgsFeatureRequest&, int&, QAtomicInt&}]’
/home/uclaros/dev/mobile/input/app/featuresmodel.cpp:46:54:   required from here
/usr/include/c++/12/type_traits:3034:11: error: no type named ‘type’ in ‘struct std::invoke_result<QList<QgsFeature> (FeaturesModel::*)(QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt&), FeaturesModel*, QPromise<void>&, QgsVectorLayerFeatureSource*, QgsFeatureRequest, int, QAtomicInt>’
 3034 |     using invoke_result_t = typename invoke_result<_Fn, _Args...>::type;
      |           ^~~~~~~~~~~~~~~
/home/uclaros/dev/mobile/input/app/featuresmodel.cpp: In member function ‘virtual void FeaturesModel::populate()’:
/home/uclaros/dev/mobile/input/app/featuresmodel.cpp:46:54: error: cannot convert ‘QFuture<void>’ to ‘const QFuture<QList<QgsFeature> >&’
   46 |     mSearchResultWatcher.setFuture( QtConcurrent::run( &FeaturesModel::fetchFeatures, this, source, req, searchId, mNextSearchId ) );
      |                                     ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                      |
      |                                                      QFuture<void>
In file included from /usr/include/x86_64-linux-gnu/qt6/QtCore/QFutureWatcher:1,
                 from /home/uclaros/dev/mobile/input/app/featuresmodel.h:14:
/usr/include/x86_64-linux-gnu/qt6/QtCore/qfuturewatcher.h:175:71: note:   initializing argument 1 of ‘void QFutureWatcher<T>::setFuture(const QFuture<T>&) [with T = QList<QgsFeature>]’
  175 | Q_INLINE_TEMPLATE void QFutureWatcher<T>::setFuture(const QFuture<T> &_future)
      |                                                     ~~~~~~~~~~~~~~~~~~^~~~~~~

```
</details>